### PR TITLE
Bench summary

### PR DESCRIPTION
**Pull request by Alizter**
We run a script at the end of a bench to sort all the lines into a file. I've set it so that the file is kept as an artifact, we can tweak it to display a nice summary if this works.

### CODE REVIEW

File: dev/bench/bench.sh [hunk 1]
```diff
@@ -572,3 +572,10 @@ $not_installable_coq_opam_packages"
 fi
 
 zulip_edit "Bench complete: all packages successfully installed."
+
+## Generate line summary
```
Comment (SkySkimmer) [inline dev/bench/bench.sh:576]: Should be done before the `if [ -n "$not_installable_coq_opam_packages" ]; then`

File: dev/bench/bench-reader.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    while read -r _ line ; do
+        read -r # empty line
+        read -r _ t1
+        read -r _ t2
+        read -r # grep's -- line
+#        line=${line%%:*}
+        f=${f#./}
+        t1=${t1%s}
+        t2=${t2%s}
+        diff=$(echo "$t2 - $t1" | jq)
```
Comment (SkySkimmer) [inline dev/bench/bench-reader.sh:29]:
```suggestion
        diff=$(jq -n "$t2 - $t1")
```
is what I suggested

**Comment by Alizter** [comment]
Nice this is working, I will add a script at the end of the job to print the head and tail of the file so we can get a preview of the results.

**Comment by Alizter** [comment]
I'd like to use the tabling from the render_results script so I've split it into some proper OCaml modules. From there I can call the table stuff. Let's see if it works.

**Comment by Alizter** [comment]
I would like to use the Table module from render_results to make the results table pretty. I had a go at separating it out into a common library, but it wasn't easy to use. At this point I am not sure it is worth it any more.

File: clib/table.ml [hunk 3]
```diff
@@ -0,0 +1,138 @@
+(************************************************************************)
```
Comment (ejgallego) [inline clib/table.ml:1]: Umm, maybe this should go into a private library for bench use only?

Comment (Alizter) [reply clib/table.ml:1]: I tried this but failed to get a shared library. Also the current ocaml scripts work out-of-the-box, I have no idea how to make shared private libraries do this.

Comment (ejgallego) [reply clib/table.ml:1]: just turn the script into an executable, and use `dune exec -- `  as we do for configure. This way we get better editor support, etc....

**Comment by Alizter** [comment]
OK I have written an executable called render_line_results that extracts the timings from the html files and summarises them near the end of our bench script.

In order to pretty print these tables, I have reused the table printer from the render_results script by putting it in its own library. render_results is now an executable rather than a toplevel script. The table library has been enhanced with some alignment fine-tuning. cc @ppedrot

I will run a full bench tonight in order to see the results in the morning.

There are still some more things we can tweak:
 * Do we want to display more or less line timings? Currently we have 25 for both the top speed ups and slow downs.
 * Do we want to dump all the timing data in a file somewhere so that it can be inspected?

**Comment by Alizter** [comment]
Here is the output of the bench:

<details>
<summary>TOP 25 SLOW DOWNS </summary>

```coq
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                       TOP 25 SLOW DOWNS                                                       │
│                                                                                                                               │
│   OLD       NEW      DIFF   %DIFF   Ln                   FILE                                                                 │
├───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ 210.1550  211.8670  1.7120   0.81%  139  coq-fiat-crypto/src/UnsaturatedSolinasHeuristics/Tests.v.html                        │
│ 120.9870  122.0630  1.0760   0.89%  153  coq-fiat-crypto/src/UnsaturatedSolinasHeuristics/Tests.v.html                        │
│ 116.2330  116.9510  0.7180   0.62%  962  coq-performance-tests-lite/src/fiat_crypto_via_setoid_rewrite_standalone.v.html      │
│ 116.0840  116.7910  0.7070   0.61%  992  coq-performance-tests-lite/src/fiat_crypto_via_setoid_rewrite_standalone.v.html      │
│  25.9110   26.4050  0.4940   1.91%   13  coq-fourcolor/theories/job231to234.v.html                                            │
│  20.9980   21.4560  0.4580   2.18%   13  coq-fourcolor/theories/job490to494.v.html                                            │
│  19.1280   19.5350  0.4070   2.13%  773  coq-category-theory/Construction/Comma/Adjunction.v.html                             │
│   8.7910    9.1860  0.3950   4.49%  297  coq-category-theory/Construction/Comma/Adjunction.v.html                             │
│   6.3650    6.7500  0.3850   6.05%  556  coq-category-theory/Construction/Comma/Adjunction.v.html                             │
│  41.7140   42.0570  0.3430   0.82%  235  coq-rewriter/src/Rewriter/Rewriter/Examples/PerfTesting/LiftLetsMap.v.html           │
│   9.4340    9.7720  0.3380   3.58%  545  coq-category-theory/Construction/Comma/Adjunction.v.html                             │
│   5.2060    5.5200  0.3140   6.03%   86  coq-engine-bench-lite/coq/PerformanceDemos/rewrite_strat_repeated_app.v.html         │
│   5.2100    5.5180  0.3080   5.91%  329  coq-category-theory/Construction/Comma/Adjunction.v.html                             │
│  18.6290   18.9170  0.2880   1.55%   78  coq-rewriter/src/Rewriter/Rewriter/Examples/PerfTesting/SieveOfEratosthenes.v.html   │
│  26.9310   27.2180  0.2870   1.07%   13  coq-fourcolor/theories/job618to622.v.html                                            │
│  26.9630   27.2420  0.2790   1.03%   13  coq-fourcolor/theories/job291to294.v.html                                            │
│  18.6150   18.8930  0.2780   1.49%  237  coq-category-theory/Construction/Comma/Adjunction.v.html                             │
│  12.7890   13.0670  0.2780   2.17%   65  coq-category-theory/Construction/Comma/Natural/Transformation.v.html                 │
│   0.8660    1.1440  0.2780  32.10%  133  coq-category-theory/Structure/Monoidal/Proofs.v.html                                 │
│   9.9020   10.1680  0.2660   2.69%  289  coq-category-theory/Construction/Comma/Adjunction.v.html                             │
│  27.3070   27.5620  0.2550   0.93%   13  coq-fourcolor/theories/job223to226.v.html                                            │
│   8.3150    8.5590  0.2440   2.93%   70  coq-category-theory/Construction/Comma/Natural/Transformation.v.html                 │
│  22.2250   22.4650  0.2400   1.08%   93  coq-category-theory/Functor/Construction/Product/Monoidal.v.html                     │
│   1.5280    1.7620  0.2340  15.31%  826  coq-unimath/UniMath/Bicategories/DisplayedBicats/Examples/MonadKtripleBiequiv.v.html │
│  33.7800   34.0080  0.2280   0.67%   13  coq-fourcolor/theories/job254to270.v.html                                            │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

</details>

<details>
<summary>TOP 25 SPEED UPS</summary>

┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                           TOP 25 SPEED UPS                                                            │
│                                                                                                                                       │
│   OLD      NEW     DIFF    %DIFF    Ln                    FILE                                                                        │
├───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ 61.9050  61.0960  -0.8090  -1.31%   137  coq-fiat-parsers/src/Parsers/Refinement/SharpenedJSON.v.html                                 │
│ 92.4800  91.6930  -0.7870  -0.85%   803  coq-vst/veric/binop_lemmas4.v.html                                                           │
│ 10.3560   9.6910  -0.6650  -6.42%   199  coq-category-theory/Instance/Adjoints.v.html                                                 │
│ 37.8850  37.4000  -0.4850  -1.28%    96  coq-vst/veric/binop_lemmas5.v.html                                                           │
│ 45.8600  45.3780  -0.4820  -1.05%    57  coq-category-theory/Construction/Comma/Natural/Transformation.v.html                         │
│ 16.3600  15.8800  -0.4800  -2.93%    90  coq-category-theory/Functor/Strong/Product.v.html                                            │
│ 23.7700  23.3090  -0.4610  -1.94%    23  coq-fiat-crypto/src/Rewriter/Passes/Arith.v.html                                             │
│ 39.2600  38.8270  -0.4330  -1.10%   633  coq-unimath/UniMath/SubstitutionSystems/Signatures.v.html                                    │
│ 57.4250  57.0280  -0.3970  -0.69%   820  coq-unimath/UniMath/SubstitutionSystems/LiftingInitial.v.html                                │
│ 18.9310  18.5790  -0.3520  -1.86%   994  coq-perennial/src/program_proof/memkv/rpc_proof.v.html                                       │
│ 51.3140  50.9780  -0.3360  -0.65%   587  coq-unimath/UniMath/SubstitutionSystems/SignatureExamples.v.html                             │
│ 12.4360  12.1120  -0.3240  -2.61%   635  coq-perennial/src/program_proof/memkv/rpc_proof.v.html                                       │
│ 22.4000  22.0800  -0.3200  -1.43%   828  coq-vst/veric/binop_lemmas2.v.html                                                           │
│ 26.3480  26.0480  -0.3000  -1.14%   510  coq-rewriter/src/Rewriter/Rewriter/Wf.v.html                                                 │
│  8.9460   8.6460  -0.3000  -3.35%    62  coq-unimath/UniMath/Bicategories/DisplayedBicats/DispBuilders.v.html                         │
│ 29.5650  29.2730  -0.2920  -0.99%    13  coq-fourcolor/theories/job399to438.v.html                                                    │
│ 12.2020  11.9140  -0.2880  -2.36%   865  coq-unimath/UniMath/Bicategories/DisplayedBicats/FiberBicategory/CodomainFiber.v.html        │
│ 17.6010  17.3430  -0.2580  -1.47%   271  coq-category-theory/Construction/Comma/Adjunction.v.html                                     │
│ 21.5910  21.3400  -0.2510  -1.16%   267  coq-perennial/src/program_proof/memkv/memkv_coord_start_proof.v.html                         │
│  8.1860   7.9480  -0.2380  -2.91%  1098  coq-category-theory/Functor/Construction/Product/Monoidal.v.html                             │
│ 12.8610  12.6310  -0.2300  -1.79%  1382  coq-perennial/src/program_proof/txn/twophase_sub_logical_reln_defs.v.html                    │
│ 31.8540  31.6250  -0.2290  -0.72%   178  coq-vst/veric/expr_lemmas4.v.html                                                            │
│  5.2940   5.0750  -0.2190  -4.14%    79  coq-category-theory/Functor/Strong/Product.v.html                                            │
│ 16.4790  16.2720  -0.2070  -1.26%   125  coq-vst/veric/binop_lemmas6.v.html                                                           │
│ 17.6690  17.4650  -0.2040  -1.15%  1886  coq-unimath/UniMath/Bicategories/MonoidalCategories/ActionBasedStrongFunctorsMonoidal.v.html │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘


</details>

<details>
<summary>Bench results</summary>

┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│        package_name         │   NEW      OLD    PDIFF │      NEW             OLD        PDIFF │      NEW             OLD        PDIFF │   NEW      OLD    PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-color │  207.20   208.77  -0.75 │   938075444575    943329605986  -0.56 │  1327625527835   1327611108192   0.00 │ 1150936  1150628   0.03 │   0    0    nan │
│                  coq-flocq3 │   76.16    76.68  -0.68 │   346675918189    348844660513  -0.62 │   456862711414    456956225896  -0.02 │  990176   994740  -0.46 │   9    0    nan │
│                     coq-vst │ 1047.60  1052.88  -0.50 │  4778561826789   4802780411155  -0.50 │  7715398357365   7715475166102  -0.00 │ 2144040  2143248   0.04 │   0    0    nan │
│                    coq-corn │  749.14   752.51  -0.45 │  3416836775255   3432173626729  -0.45 │  5276812687389   5276768527773   0.00 │  879356   877932   0.16 │   0    0    nan │
│                 coq-unimath │ 4250.77  4269.42  -0.44 │ 19449721096814  19533730563827  -0.43 │ 38989288523327  38989680782694  -0.00 │ 3270848  3272744  -0.06 │   0    0    nan │
│            coq-fiat-parsers │  331.39   332.46  -0.32 │  1491436892394   1496726078742  -0.35 │  2453483206554   2453323492986   0.01 │ 2697820  2696720   0.04 │   0    0    nan │
│         coq-metacoq-erasure │   63.36    63.56  -0.31 │   284169746160    284696499004  -0.19 │   433280888040    433446280261  -0.04 │ 1110452  1113588  -0.28 │   0    0    nan │
│          coq-mathcomp-field │   97.98    98.27  -0.30 │   448695855819    450073415502  -0.31 │   739006693433    738974863241   0.00 │  703844   706560  -0.38 │   0    0    nan │
│                    coq-hott │  161.80   162.27  -0.29 │   735144548096    736129971701  -0.13 │  1160388350661   1160360842753   0.00 │  617440   617284   0.03 │   0    0    nan │
│    coq-metacoq-translations │   19.92    19.97  -0.25 │    89714802689     89725112601  -0.01 │   156889889270    156907483055  -0.01 │  971496   970932   0.06 │   0    0    nan │
│        coq-metacoq-template │   76.40    76.55  -0.20 │   341660025553    342005733269  -0.10 │   513820322090    513948358438  -0.02 │  911128   913132  -0.22 │   0    0    nan │
│               coq-perennial │ 3954.65  3962.01  -0.19 │ 18058334404747  18085353640869  -0.15 │ 29760447700650  29760048839500   0.00 │ 2626616  2626680  -0.00 │   0    0    nan │
│                 coq-bignums │   27.53    27.57  -0.15 │   125356841902    125602788744  -0.20 │   171949821665    171929719778   0.01 │  471440   468520   0.62 │   0    0    nan │
│              coq-verdi-raft │  556.90   557.63  -0.13 │  2540084926632   2543981805008  -0.15 │  3927494110133   3927315451326   0.00 │ 1261400  1261068   0.03 │   0    0    nan │
│           coq-iris-examples │  412.23   412.75  -0.13 │  1876275378911   1879823903899  -0.19 │  2839029548866   2839082984665  -0.00 │ 1185104  1184604   0.04 │   0    0    nan │
│              coq-coquelicot │   33.24    33.28  -0.12 │   149619693732    149826676175  -0.14 │   196424329994    196423928277   0.00 │  756280   758232  -0.26 │   0    0    nan │
│  coq-performance-tests-lite │  772.14   773.01  -0.11 │  3507144974846   3509475913224  -0.07 │  5942817098206   5943146377997  -0.01 │ 1824528  1824928  -0.02 │   3    0    nan │
│       coq-mathcomp-solvable │   79.54    79.60  -0.08 │   363612610843    364060912129  -0.12 │   555327881224    555365874358  -0.01 │  702672   700020   0.38 │   0    0    nan │
│        coq-mathcomp-algebra │   58.11    58.13  -0.03 │   265616829045    265794588012  -0.07 │   361773832464    361722953829   0.01 │  599456   595276   0.70 │   0    0    nan │
│             coq-fiat-crypto │ 2787.35  2787.90  -0.02 │ 12679929083447  12681369345301  -0.01 │ 23240895656948  23240597460793   0.00 │ 3851448  3851440   0.00 │   1    0    nan │
│                  coq-geocoq │  690.63   690.70  -0.01 │  3149977178802   3149398032400   0.02 │  4989819092215   4989752594412   0.00 │ 1076876  1076284   0.06 │   0    0    nan │
│            coq-math-classes │   89.95    89.95   0.00 │   409121077214    408005505890   0.27 │   571445784301    571452013908  -0.00 │  502512   502636  -0.02 │   0    0    nan │
│      coq-mathcomp-character │   66.76    66.75   0.01 │   305665807408    305214301493   0.15 │   463940905353    463963970474  -0.00 │  829708   829724  -0.00 │   0    0    nan │
│         coq-category-theory │ 1623.80  1623.09   0.04 │  7433624880216   7431353466042   0.03 │ 15548503840106  15548483683790   0.00 │ 1469900  1471252  -0.09 │   5    0    nan │
│      coq-mathcomp-odd-order │  523.39   523.14   0.05 │  2398344925267   2396634271260   0.07 │  4065772214239   4065749140065   0.00 │ 1207956  1209908  -0.16 │   0    0    nan │
│                 coq-coqutil │   34.73    34.71   0.06 │   155552154986    155436507772   0.07 │   209851018869    209851900147  -0.00 │  607800   607568   0.04 │   0    0    nan │
│                coq-coqprime │   42.51    42.48   0.07 │   192626159146    192690458869  -0.03 │   283392132751    283345758957   0.02 │  762140   764796  -0.35 │   0    0    nan │
│               coq-fiat-core │   55.10    55.06   0.07 │   237182313657    237320752525  -0.06 │   332268727927    332261223795   0.00 │  479968   481328  -0.28 │   1    0    nan │
│               coq-fourcolor │ 1456.24  1454.95   0.09 │  6645041814350   6635884397776   0.14 │ 12066612337631  12066396455892   0.00 │  758424   759832  -0.19 │   0    0    nan │
│       coq-mathcomp-fingroup │   21.39    21.37   0.09 │    97744416241     97441549854   0.31 │   140302594981    140306940035  -0.00 │  488088   483796   0.89 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │  925.40   924.31   0.12 │  4222077346512   4216841865672   0.12 │  7267197678807   7268736294492  -0.02 │ 1146276  1147048  -0.07 │   0    0    nan │
│                    coq-core │  100.33   100.19   0.14 │   407887098613    402637106550   1.30 │   441796800978    441886658926  -0.02 │  277468   276560   0.33 │   0    0    nan │
│     coq-metacoq-safechecker │  195.37   194.87   0.26 │   889128444191    887198911372   0.22 │  1203010811947   1202866403336   0.01 │ 1232540  1232764  -0.02 │   0    0    nan │
│                coq-rewriter │  371.50   370.32   0.32 │  1692801354934   1685408025677   0.44 │  2772671589067   2772602996876   0.00 │ 1064236  1060416   0.36 │   0    0    nan │
│      coq-mathcomp-ssreflect │   24.84    24.76   0.32 │   113059852021    112823030046   0.21 │   140553282788    140537073410   0.01 │  538652   535912   0.51 │   0    0    nan │
│                coq-bedrock2 │  218.47   217.74   0.34 │   997091431105    992645521925   0.45 │  1795138908003   1795046742579   0.01 │ 2051980  2051992  -0.00 │   1    0    nan │
│           coq-metacoq-pcuic │  710.89   707.99   0.41 │  3211748291165   3197727089252   0.44 │  4660683355245   4661061426158  -0.01 │ 1663960  1661592   0.14 │   0    0    nan │
│                   coq-verdi │   47.31    47.10   0.45 │   214832722773    214158064691   0.32 │   319030429401    318990375969   0.01 │  839660   841812  -0.26 │   0    0    nan │
│       coq-engine-bench-lite │  176.31   175.16   0.66 │   763879583683    759200846229   0.62 │  1490406746436   1489597873924   0.05 │ 1234572  1235004  -0.03 │   0    0    nan │
│                coq-compcert │  280.04   277.66   0.86 │  1270848138861   1261296433774   0.76 │  1889279694024   1889270473127   0.00 │ 1110180  1106972   0.29 │   0    0    nan │
│                  coq-stdlib │  399.99   395.48   1.14 │  1662656838597   1640819831906   1.33 │  1417081628114   1417194915398  -0.01 │  609320   612824  -0.57 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘

</details>

Note that the header alignments in the bench results have changed from right to mid. We can easily revert this change if needed.

File: dev/bench/bench.sh [hunk 4]
```diff
@@ -535,6 +537,12 @@ du -ha "$working_dir" > "$working_dir/files.listing"
 #
 # The next script processes all these files and prints results in a table.
 
+
+# Print line by line slow downs and speed ups
+cd "$working_dir/html"
+rendered_line_results=$($render_line_results)
+echo "$rendered_line_results"
```
Comment (SkySkimmer) [inline dev/bench/bench.sh:544]:
```suggestion
$render_line_results
```

**Comment by SkySkimmer** [comment]
>Do we want to dump all the timing data in a file somewhere so that it can be inspected?

I think that would be nice.

**Comment by Alizter** [comment]
Alright all our tables (speedups, slowdowns, all timings and summary) are dumped into the timings folder which should be a browsable artifact.

**Comment by Alizter** [comment]
Quickly testing the artifacts then I will remove the commit.

**Comment by Alizter** [comment]
Here are the artifacts working. Currently set up only to test one project, I will remove that change now.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**